master/buildbot/db/migrate/versions/052_cascading_set_null.py

```diff
@@ -0,0 +1,90 @@
+# This file is part of Buildbot.  Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+from __future__ import absolute_import
+from __future__ import print_function
+from future.utils import iteritems
+
+import sqlalchemy as sa
+from migrate.changeset.constraint import ForeignKeyConstraint
+from migrate.exceptions import NotSupportedError
+
+from buildbot.util import sautils
+
+
+def upgrade(migrate_engine):
+    metadata = sa.MetaData()
+    metadata.bind = migrate_engine
+
+    table_names = set(TABLES_FKEYS_SET_NULL.keys())
+    table_names.update(TABLES_COLUMNS_NOT_NULL.keys())
+
+    tables = {}
+    for t in table_names:
+        tables[t] = sautils.Table(t, metadata, autoload=True)
+
+    fks_to_change = []
+    # We need to parse the reflected model in order to find the automatic
+    # fk name that was put.
+    # Mysql and postgres have different naming convention so this is not very
+    # easy to have generic code working.
+    for t, table in iteritems(tables):
+        if t in TABLES_FKEYS_SET_NULL:
+            keys = TABLES_FKEYS_SET_NULL[t]
+            for fk in table.constraints:
+                if not isinstance(fk, sa.ForeignKeyConstraint):
+                    continue
+                for c in fk.elements:
+                    if str(c.column) in keys:
+                        # migrate.xx.ForeignKeyConstraint is changing the model
+                        # so initializing here would break the iteration
+                        # (Set changed size during iteration)
+                        fks_to_change.append((
+                            table, (fk.columns, [c.column]),
+                            dict(name=fk.name, ondelete='SET NULL')))
+        if t in TABLES_COLUMNS_NOT_NULL:
+            cols = TABLES_COLUMNS_NOT_NULL[t]
+            for c in table.columns:
+                if c.name in cols:
+                    c.alter(nullable=False)
+
+    for table, args, kwargs in fks_to_change:
+        fk = ForeignKeyConstraint(*args, **kwargs)
+        table.append_constraint(fk)
+        try:
+            fk.drop()
+        except NotSupportedError:
+            # some versions of sqlite do not support drop,
+            # but will still update the fk
+            pass
+        fk.create()
+
+
+TABLES_FKEYS_SET_NULL = {
```

Comment:
I don't see parent change ids in this list.

I am not sure how to read those tables.

Will the buildrequest_claims associated with a masterid be deleted when I delete that master?

I expect so, but reading the code it looks like it will just nullify the FK

Reply:
Yes the parent change ids is:

```python
```

And yes, when deleting a master, the related `buildrequest_claims.masterid` column will be set to `NULL`. This is one of the cases where I was not sure what to do.

Maybe the correct solution is to forbit `buildrequest_claims.masterid` to be `NULL`. So move it from `TABLES_FKEYS_SET_NULL` to `TABLES_COLUMNS_NOT_NULL`.

Reply:
Your reasoning here is that CASCADE_DELETE and NOT_NULL are equivalent. I disagree with this.

There are cases where you want to cascade delete even if the FK can be null.

if you delete a patch, you want to cascade delete the sourcestamps, and then the builds associated with it. because the build makes no sense without the patch.
But.. a sourcestamp can be made without patch, and this is meaningful.

in this list, I think only  following make sense
```
   'buildsets': ['parent_buildid'],
    'changes': ['changes.changeid'],
```
The other shall be ON DELETE CASCADE and NULLABLE imho

Reply:
Ok. But when you delete a worker, do you want to delete all its builds?

Reply:
I think a build without worker makes no sense indeed. It should not be NULLABLE.
It might be NULLABLE temporarily for technical reasons

Reply:
What should we do about this comment then?

```python
        # worker which performed this build
        # keep nullable to support worker-free builds
```

Reply:
I went back in time as far as I could, the comment was added here:

https://github.com/buildbot/buildbot/commit/9dbb106fc2b274527a08b1b945aa0897a471cee3#diff-58b3908894b0546af761475a0243ffd1R98

With no specific reason. By looking at the code that creates builds. There is no way we can have builds without any worker. I'll change `builds.workerid` to `nullable=False` and leave `ondelete='CASCADE'`.